registry/consul.go

```diff
@@ -0,0 +1,205 @@
+//
+// Copyright (c) 2018
+// IOTech
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package registry
+
+import (
+	"errors"
+	"reflect"
+	"strconv"
+	"strings"
+
+	"fmt"
+	consulapi "github.com/hashicorp/consul/api"
+)
+
+type ConsulClient struct {
+	Consul *consulapi.Client
+}
+
+func (c *ConsulClient) Init(config Config) error {
+	var err error // Declare error to be used throughout function
+
+	// Connect to the Consul Agent
+	defaultConfig := &consulapi.Config{}
+	defaultConfig.Address = config.Address + ":" + strconv.Itoa(config.Port)
+	c.Consul, err = consulapi.NewClient(defaultConfig)
+	if err != nil {
+		return err
+	}
+
+	// Register the Service
+	fmt.Println("Register the Service ...")
+	err = c.Consul.Agent().ServiceRegister(&consulapi.AgentServiceRegistration{
+		Name:    config.ServiceName,
+		Address: config.ServiceAddress,
+		Port:    config.ServicePort,
+	})
+	if err != nil {
+		return err
+	}
+
+	// Register the Health Check
+	fmt.Println("Register the Health Check ...")
+	err = c.Consul.Agent().CheckRegister(&consulapi.AgentCheckRegistration{
+		Name:      "Health Check: " + config.ServiceName,
+		Notes:     "Check the health of the API",
+		ServiceID: config.ServiceName,
+		AgentServiceCheck: consulapi.AgentServiceCheck{
+			HTTP:     config.CheckAddress,
+			Interval: config.CheckInterval,
+		},
+	})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (c *ConsulClient) GetServiceEndpoint(serviceKey string) (ServiceEndpoint, error) {
+	services, err := c.Consul.Agent().Services()
+	if err != nil {
+		return ServiceEndpoint{}, err
+	}
+
+	endpoint := ServiceEndpoint{}
+	for key, service := range services {
+		if key == serviceKey {
+			endpoint.Port = service.Port
+			endpoint.Key = key
+			endpoint.Address = service.Address
+		}
+	}
+	return endpoint, nil
+}
+
+func (c *ConsulClient) CheckKeyValuePairs(configuration interface{}, applicationName string, profiles []string) error {
```

Comment:
did you check it? if this one copy/paste from edgex-go I'm afraid it will not work properly with device configuration structure.

In edgex-go configs have flat structure like:
checkUrl ="bla"
port = 8000
serviceName = ""

in device-sdk we had groups like:
[Service]
  Name = "bla"

So algorithm for parsing from edgex-go doesn't work properly for device-sdk

Reply:
Sure. So maybe just better to avoid functionality 'CheckKeyValuePairs' in this commit and provide it separately?

Reply:
but is it necessary to fix the issue? let's keep commits for pull request modular. If this is not a root of problem we should avoid such change in fix commit/PR. And add it in separate. If commit describe that it should be a fix, we provide a fix - nothing more. Also we already have issue for stroing/get config in Consul https://github.com/edgexfoundry/device-sdk-go/issues/6

Reply:
Ok, I have refactor GetConsulClient function. Only check consul available and return consulClient.